Update (#1)
DevExpress-Examples/asp-net-mvc-spreadsheet-work-with-database: Readme.md

[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T190813)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->
# Spreadsheet for ASP.NET MVC - How to save and load documents from a database

> This example demonstrates how to configure the [Spreadsheet extension](https://docs.devexpress.com/AspNetMvc/17113/components/spreadsheet) to save and load documents from a database.

Reviewer: A potential issue here: you say "to save and **load** documents **from** a database"

Thing is, that you can **load** documents **from** a database. At the same time, you can **save** documents **to** a database.

That is, two different prepositions designate two different information flow directions. How can we resolve this issue?

1. The simplest way is to join "save/load" and "to/from" through a slash character.

```suggestion
This example demonstrates how to configure the [Spreadsheet extension](https://docs.devexpress.com/AspNetMvc/17113/components/spreadsheet) to save/load documents to/from a database.
```

OR

2. Alternatively, we can find a synonym that can logically generalize the both operations. That would be "store" or "persist" in this scenario. If you find "to store documents in a database" too abstract, you may stick to the solution with slash characters. However, I don't see any fundamental issues in the second scenario.

```suggestion
This example demonstrates how to configure the [Spreadsheet extension](https://docs.devexpress.com/AspNetMvc/17113/components/spreadsheet) to store documents in a database.
```

I'm OK with either variant. However, you may come up with a better variant. Who knows.)

![Connect Spreadsheet to Database](connect-spreadsheet-to-database.png)

## Overview

Follow the steps below to configure the Spreadsheet extension to work with a database:

> 1. Create a class that models a spreadsheet document. This class should be able to store a document's identifier, format, and content:

Reviewer 2: ```suggestion
1. Create a class that models a spreadsheet document. This class should be able to store document identifiers, formatting, and content:
```

Author: I'm not sure we could replace the "format" word with the "formatting" word. In this sentence I'm talking about document file formats (such as DOCX, PDF, TXT). Document formatting term refers to the way a document looks and addresses things like margins, spacing, font size.

Can we say something like the following?
```suggestion
1. Create a class that models a spreadsheet document. This class should be able to store document identifiers, file formats, and content:
```

Reviewer 2: this works

        protected override void OnModelCreating(DbModelBuilder modelBuilder) {
            throw new UnintentionalCodeFirstException();
        }
        public virtual DbSet<Doc> Docs { get; set; }
    }

    public partial class Doc {
        public int Id { get; set; }
        public byte[] DocBytes { get; set; }
        public string Comments { get; set; }
    }
    ```

3. Add connection strings to the **Web.config** file:

    ```xml
    <connectionStrings>
        <add name="DocumentsConnectionString" connectionString="Data Source=(local);Initial Catalog = Documents;User Id=sa; Password=dx;Connect Timeout=30" providerName="System.Data.SqlClient" />
        <add name="DocumentsEntities" connectionString="metadata=res://*/Models.Model1.csdl|res://*/Models.Model1.ssdl|res://*/Models.Model1.msl;provider=System.Data.SqlClient;provider connection string=&quot;Data Source=(localdb)\MSSQLLocalDB;AttachDbFilename=|DataDirectory|Documents.mdf;Integrated Security=True;connect timeout=30;MultipleActiveResultSets=True;App=EntityFramework&quot;" providerName="System.Data.EntityClient" />
    </connectionStrings>
    ```

> 4. Create a helper class that works with the database. The class should be able to get a document from a database as an array of bytes and save a byte array back to the database:

Reviewer: > The class should be able to get a document from a database as an array of bytes and save **a** byte array **back** to the database:

**A potential issue**. At first, everything looks the way you describe the abstract functionality of your helper class. The class should be able to load a document as an array of bytes and save a byte array to the database.

However, you said "back" in "save a byte array **back** to the database". If we save something "back", it implies that we speak of the same array of bytes, or the same document. In this case, a definite article is in order. Or getting rid of "back" as an indication that implies definiteness in this particular usage scenario.

```suggestion
4. Create a helper class that works with the database. The class should be able to load a document from a database as an array of bytes and save the byte array back to the database:
```

OR

```suggestion
4. Create a helper class that works with the database. The class should be able to load a document from a database as an array of bytes and save a document as a byte array to the database:
```

Again, I'm OK with either variant, however, you may come up with something better.

P.S. Why "get" a document rather than "load"? "Load" is the opposed operation to "Save" after all.

    ```cs
    public class DataHelper {
        public static byte[] GetDocument() {
            DocumentsEntities context = new DocumentsEntities();
            return context.Docs.FirstOrDefault().DocBytes.ToArray();
        }

        public static void SaveDocument(byte[] bytes) {
            DocumentsEntities context = new DocumentsEntities();
            context.Docs.FirstOrDefault().DocBytes = bytes;
            context.SaveChanges();
        }
    }
    ```

> 5. Create the `SpreadsheetSettingsHelper` helper class that configures and returns the [SpreadsheetSettings](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetSettings?p=netframework) object. Specify the object's [Name](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SettingsBase.Name) and [CallbackRouteValues](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetSettings.CallbackRouteValues) properties. Set the [Saving](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetSettings.Saving?p=netframework) property to a function that converts an open document to a byte array and saves it to the database:

Reviewer: ```suggestion
5. Create the `SpreadsheetSettingsHelper` helper class that configures and returns the [SpreadsheetSettings](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetSettings?p=netframework) object. Specify the object's [Name](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SettingsBase.Name) and [CallbackRouteValues](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetSettings.CallbackRouteValues) properties. Set the [Saving](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetSettings.Saving?p=netframework) property to a function that converts an opened document to a byte array and saves it to the database:
```

            settings.Saving = (s, e) => {
                byte[] docBytes = SpreadsheetExtension.SaveCopy("SpreadsheetName", DocumentFormat.Xlsx);
                DataHelper.SaveDocument(docBytes);
                e.Handled = true;
            };
            // ...
            return settings;
        }
    }
    ```
    
6. In a partial view, pass the settings that the `SpreadsheetSettingsHelper` class configures to the [Spreadsheet](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UI.ExtensionsFactory.Spreadsheet(DevExpress.Web.Mvc.SpreadsheetSettings)?p=netframework) extension method to create the Spreadsheet. Call the Spreadsheet's [Open](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.SpreadsheetExtension.Open(System.String-DevExpress.Spreadsheet.DocumentFormat-System.Func-System.Byte---)?p=netframework) method to open the document stored in the model:

    ```razor
    @model DXWebApplication23.Models.SpreadsheetData

    @Html.DevExpress().Spreadsheet(SpreadsheetSettingsHelper.GetSpreadsheetSettings()).Open(
        Model.DocumentId, 
        Model.DocumentFormat, 
        () => { return Model.Document; }
    ).GetHtml()
    ```

> 7. Configure the controller. In the `Index` action method, get a document from the database and save it to a model. In the action method you assigned to the `CallbackRouteValues` setting, use the `SpreadsheetSettingsHelper` class to configure Spreadsheet settings, then return them back to the client: 

Reviewer: ```suggestion
7. Configure the controller. In the `Index` action method, load a document from the database and save it to a model. In the action method you assigned to the `CallbackRouteValues` setting, use the `SpreadsheetSettingsHelper` class to configure Spreadsheet settings, then return them back to the client: 
```

            };
            return View(model);
        }

        // The method assigned to the Spreadsheet's CallbackRouteAction property in the helper class
        public ActionResult SpreadsheetPartial() { 
            return SpreadsheetExtension.GetCallbackResult(SpreadsheetSettingsHelper.GetSpreadsheetSettings());
        }
    }
    ```

## Files to Review

* [SpreadsheetPartial.cshtml](./CS/DXWebApplication23/Views/Home/SpreadsheetPartial.cshtml)
* [HomeController.cs](./CS/DXWebApplication23/Controllers/HomeController.cs) (VB: [HomeController.vb](./VB/DXWebApplication23/Controllers/HomeController.vb))
* [DataHelper.cs](./CS/DXWebApplication23/Models/DataHelper.cs) (VB: [DataHelper.vb](./VB/DXWebApplication23/Models/DataHelper.vb))
* [Index.cshtml](./CS/DXWebApplication23/Views/Home/Index.cshtml)

## Documentation

* [Creating a Connection String and Working with SQL Server LocalDB](https://learn.microsoft.com/en-us/aspnet/mvc/overview/getting-started/introduction/creating-a-connection-string)
* [Office Document Management](https://docs.devexpress.com/AspNetMvc/402337/common-features/office-document-management)

## More Examples

> * [Spreadsheet for ASP.NET Web Forms - How to save and load documents from a database](https://github.com/DevExpress-Examples/aspxspreadsheet-how-to-save-and-load-documents-from-a-database-t190812)

Reviewer: This topic has the same potential issue regarding the "save" and "load" generalization under the "from" preposition.
